izer/assets.py

```diff
@@ -112,3 +112,162 @@ def from_template(
+            # RISC-V projects will look for ...-combined.elf
+            elf_file = "${config:proj_name}-combined.elf"
+        else:
+            # Default is project name (which defaults to folder name in template)
+            elf_file = "${config:proj_name}.elf"
+
+    if defines == "":
+        defines = state.defines
+
+    if i_paths == "":
+        i_paths = state.eclipse_includes  # TODO: rename state var to 'includes'?
+
+    if v_paths == "":
+        pass  # TODO: Support adding browse paths on command line
+    # ---
+
+    tmp = []  # Work-horse list, linter be nice
+    # Parse defines...
+    # ---
+    tmp = defines.split(" ")
+
+    if state.defines_arm != "":
+        # Split & append Arm defines
+        for d in state.defines_arm.split(" "):
+            tmp.append(d)
+
+    if state.riscv and state.defines_riscv != "":
+        # Split & append risc-v defines
+        for d in state.defines_arm.split(" "):
+            tmp.append(d)
+
+    tmp = list(map(lambda s: s.strip("-D"), tmp))  # VS Code doesn't want -D
+    tmp = list(map("\"{0}\"".format, tmp))  # Surround with quotes
+    defines_parsed = ",\n\t\t\t\t".join(tmp)  # csv, newline, and tab alignment
+    # ---
+
+    # Parse include paths...
+    tmp = i_paths.split(" ")  # Space-separated
+    tmp = list(map("\"{0}\"".format, tmp))  # Surround with quotes
+    i_paths_parsed = ",\n\t\t\t\t".join(tmp)  # csv, newline, and tab alignment
+
+    # Parse browse paths...
+    tmp = v_paths.split(" ")  # Space-separated
+    tmp = list(map("\"{0}\"".format, tmp))  # Surround with quotes
+    v_paths_parsed = ",\n\t\t\t\t\t".join(tmp)  # csv, newline, and tab alignment
+
+    # Create template...
+    for directory, _, files in sorted(os.walk(template_dir)):
+        # ^ For each directory in the directory tree rooted at top (including top itself,
+        # but excluding '.' and '..'), yields a 3-tuple (dirpath, dirnames, filenames)
+
+        # Get current directory relative to root
+        rel_dir = os.path.relpath(directory, template_dir)
+
+        # Figure out whether we're in a subfolder of the template directory,
+        # and form output path accordingly.
+        out_path = ""
+        if rel_dir != '.':
+            # We're in a sub-folder.  Replicate this folder in the output directory
+            out_path = os.path.join(os.path.join(out_stem, out_branch), rel_dir)
+            os.makedirs(out_path, exist_ok=True)
+        else:
+            # We're in the root template folder.
+            out_path = os.path.join(out_stem, out_branch)
+
+        # Any files to copy?
+        for file in sorted(files):
+
+            if file.startswith(template_prefix):
+
+                # There is a template file to copy.  Perform string substitution in output file.
+                out_loc = os.path.join(out_path, file[len(template_prefix):])
+                with open(os.path.join(directory, file)) as in_file, \
+                        open(out_loc, "w+") as out_file:
```

Comment:
need to add encoding and mode to all open() calls.
The first one should become open(os.path.join(directory, file), mode='r', encoding='utf-8')
and the second one open(out_loc, mode='w+', encoding='utf-8')

Reply:
Added